src/libstd/sys/common/net.rs

```diff
@@ -164,6 +164,92 @@ pub fn lookup_addr(addr: &IpAddr) -> io::Result<String> {
 }
 
 ////////////////////////////////////////////////////////////////////////////////
+// Timeouts
+////////////////////////////////////////////////////////////////////////////////
+
+#[cfg(target_os = "windows")]
+fn set_timeout(socket: &Socket, dur: Option<Duration>, kind: libc::c_int) -> io::Result<()> {
+    let timeout = match dur {
+        Some(dur) => {
+            if dur.secs() == 0 && dur.extra_nanos() == 0 {
+                return Err(io::Error::new(io::ErrorKind::InvalidInput,
+                                          "cannot set a 0 duration timeout"));
+            }
+
+            let mut timeout = if dur.secs() > (libc::DWORD::max_value() / 1000) as u64 {
+                libc::DWORD::max_value()
+            } else {
+                (dur.secs() * 1000) as libc::DWORD
+            };
+            timeout = timeout.saturating_add((dur.extra_nanos() / 1000000) as libc::DWORD);
+            if timeout == 0 {
+                timeout = 1;
+            }
+            timeout
+        }
+        None => 0
+    };
+    setsockopt(socket, libc::SOL_SOCKET, kind, timeout)
+}
+
+#[cfg(not(target_os = "windows"))]
+fn set_timeout(socket: &Socket, dur: Option<Duration>, kind: libc::c_int) -> io::Result<()> {
+    let timeout = match dur {
+        Some(dur) => {
+            if dur.secs() == 0 && dur.extra_nanos() == 0 {
+                return Err(io::Error::new(io::ErrorKind::InvalidInput,
+                                          "cannot set a 0 duration timeout"));
+            }
+
+            let secs = if dur.secs() > libc::time_t::max_value() as u64 {
```

Comment:
This looks pretty similar to some code in `Condvar::wait_timeout`, perhaps those two paths could be unified in converting a `Duration` to a `libc::timeval`? If the other code is a `timespec` though ah well.